sklearn/metrics/_regression.py

```diff
@@ -1002,17 +1011,31 @@ def r2_score(
     else:
         weight = 1.0
 
-    numerator = (weight * (y_true - y_pred) ** 2).sum(axis=0, dtype=np.float64)
-    denominator = (
-        weight * (y_true - np.average(y_true, axis=0, weights=sample_weight)) ** 2
-    ).sum(axis=0, dtype=np.float64)
+    weighted = weight * (y_true - y_pred) ** 2
+    weighted_difference = weight * (
+        y_true - _average(y_true, axis=0, weights=sample_weight, xp=xp)
+    )
+
+    # We move to cpu device ahead of time since certain devices may not support
+    # float64, but we want the same precision for all devices and namespaces.
+    # This works for all protocols, except for CuPy which does not support
+    # moving data to the CPU using xp.asarray(..., device="cpu").
```

Comment:
The Array API does not specify what you can/can't pass as value for `device`. This means that we can't pass `device="cpu"` here. For more details see https://data-apis.org/array-api/2022.12/design_topics/device_support.html#syntax-for-device-assignment, in particular the "Note" in that section. Another thing that makes this tricky is that you can be "Array API compliant" and not have a "CPU like" device. An example of this is CuPY which only deals with GPUs, it doesn't have a CPU mode.

A relevant PR is #27232 where we have a similar case of a hard coded `device=`. There I introduce a "highest precision float type" function so that we can keep computation on the device, with the highest precision that is available for that library/device.

Reply:
@betatim I cherry-picked and implemented your function from #27232, works like charm! :)